sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/multi/MultiServiceManager.java

```diff
@@ -0,0 +1,282 @@
+     *   a. Upstream builds a new foo using SchedulerBuilder, which internally finds the uninstall bit and returns a
+     *      new UninstallScheduler
+     *   b. putService(foo) is called with the UninstallScheduler
+     *   c. The UninstallScheduler resumes cleanup from where it left off...
+     * 5. Sometime after the UninstallScheduler finishes cleanup, it returns FINISHED in response to offers.
+     * 6. We remove the service and invoke uninstallCallback.uninstalled(), telling upstream that it's gone. If upstream
+     *    invokes putService(foo) again at this point, the service will be relaunched from scratch because the uninstall
+     *    bit in ZK will have been cleared.
+     */
+    public void uninstallServices(Collection<String> finishedServiceNames) {
+        rwlock.lock();
+        try {
+            LOGGER.info("Marking services as uninstalling: {} (out of {} services)",
+                    finishedServiceNames, services.size());
+
+            for (String name : finishedServiceNames) {
+                AbstractScheduler currentService = services.get(name);
+                if (currentService == null) {
+                    LOGGER.warn("Service '{}' does not exist, cannot trigger uninstall", name);
+                    continue;
+                }
+                if (currentService instanceof UninstallScheduler) {
+                    // Already uninstalling
+                    LOGGER.warn("Service '{}' is already uninstalling, leaving as-is", name);
+                    continue;
+                }
+
+                // Convert the DefaultScheduler to an UninstallScheduler. It will automatically flag itself with an
+                // uninstall bit in its state store and then proceed with the uninstall. When the uninstall has
+                // completed, it will return FINISHED to its next offers() call, at which point we will remove it. If
+                // the scheduler process is restarted before uninstall has completed, the caller should have added it
+                // back via putService(). When it's added back, it should be have already been converted to an
+                // UninstallScheduler. See SchedulerBuilder.
+                AbstractScheduler uninstallScheduler = ((DefaultScheduler) currentService).toUninstallScheduler();
+                if (isRegistered) {
+                    // Don't forget to also initialize the new scheduler if relevant...
+                    uninstallScheduler.registered(false);
+                }
+                services.put(name, uninstallScheduler);
+            }
+        } finally {
+            rwlock.unlock();
+        }
+    }
+
+    /**
+     * Removes the specified services after they have completed uninstall. Any unknown service names are ignored.
+     *
+     * @return the number of services which are still present after this removal
+     */
+    public int removeServices(Collection<String> uninstalledServiceNames) {
```

Comment:
Having this return value while perhaps convenient seems weird.  I'd just have `getServicesCount` method somewhere and keep it separate.

Reply:
Fixed in upcoming PR.